packages/dds/tree/src/feature-libraries/modular-schema/modularChangeFamily.ts

```diff
@@ -92,44 +93,77 @@ export class ModularChangeFamily
         return { fieldKind, changesets: normalizedChanges };
     }
 
-    compose(changes: FieldChangeMap[]): FieldChangeMap {
+    compose(changes: TaggedChange<FieldChangeMap>[]): FieldChangeMap {
         if (changes.length === 1) {
-            return changes[0];
+            return changes[0].change;
         }
 
         const fieldChanges = new Map<FieldKey, FieldChange[]>();
         for (const change of changes) {
-            for (const [key, fieldChange] of change) {
-                getOrAddEmptyToMap(fieldChanges, key).push(fieldChange);
+            for (const [key, fieldChange] of change.change) {
+                const fieldChangeToCompose =
+                    fieldChange.revision !== undefined || change.revision === undefined
+                        ? fieldChange
+                        : {
+                              ...fieldChange,
+                              revision: change.revision,
+                              isInverse: change.isInverse,
+                          };
+
+                getOrAddEmptyToMap(fieldChanges, key).push(fieldChangeToCompose);
             }
         }
 
         const composedFields: FieldChangeMap = new Map();
         for (const [field, changesForField] of fieldChanges) {
-            const { fieldKind, changesets } = this.normalizeFieldChanges(changesForField);
-            const composedField = fieldKind.changeHandler.rebaser.compose(
-                changesets,
-                this.childComposer,
-            );
+            let composedField: FieldChange;
+            if (changesForField.length === 1) {
+                composedField = changesForField[0];
+            } else {
+                const { fieldKind, changesets } = this.normalizeFieldChanges(changesForField);
+                assert(
+                    changesets.length === changesForField.length,
+                    "Number of changes should be constant when normalizing",
+                );
+                const taggedChangesets: TaggedChange<FieldChangeset>[] = [];
+                for (let i = 0; i < changesets.length; i++) {
+                    const taggedChange: TaggedChange<FieldChangeset> = {
+                        revision: changesForField[i].revision,
+                        isInverse: changesForField[i].isInverse,
+                        change: changesets[i],
+                    };
+
+                    taggedChangesets.push(taggedChange);
+                }
```

Comment:
Consider changing `normaizeFieldChanges` to return a `{ fieldKind: FieldKind; changes: FieldChange[]; }` that preserves this information instead.